x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/MultiTypeEsField.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.type;
+
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.core.type.EsField;
+import org.elasticsearch.xpack.esql.core.type.InvalidMappedField;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * During IndexResolution it could occur that the same field is mapped to different types in different indices.
+ * The class MultiTypeEfField.UnresolvedField holds that information and allows for later resolution of the field
+ * to a single type during LogicalPlanOptimization.
+ * If the plan contains conversion expressions for the different types, the resolution will be done using the conversion expressions,
+ * in which case a MultiTypeEsField will be created to encapsulate the type resolution capabilities.
+ * This class can be communicated to the data nodes and used during physical planning to influence field extraction so that
+ * type conversion is done at the data node level.
+ */
+public class MultiTypeEsField extends EsField {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(
+        EsField.class,
+        "MultiTypeEsField",
+        MultiTypeEsField::new
+    );
+
+    private final Map<String, Expression> indexToConversionExpressions;
+
+    public MultiTypeEsField(String name, DataType dataType, boolean aggregatable, Map<String, Expression> indexToConversionExpressions) {
+        super(name, dataType, Map.of(), aggregatable);
+        this.indexToConversionExpressions = indexToConversionExpressions;
+    }
+
+    public MultiTypeEsField(StreamInput in) throws IOException {
+        // TODO: Change the conversion expression serialization to i.readNamedWriteable(Expression.class) once Expression is fully supported
+        this(
+            in.readString(),
+            DataType.readFrom(in),
+            in.readBoolean(),
+            in.readImmutableMap(StreamInput::readString, i -> ((PlanStreamInput) i).readExpression())
```

Comment:
I tend to just do `readMap( i -> ((PlanStreamInput) i).readExpression())`

Reply:
That's a little more convenient and I just trust that we don't mutate the map.

Though I should probably do:
```
readImutableMap(i -> ((PlanStreamInput) i).readExpression())
```